### ISSUE TITLE
Pry for Crystal

### ISSUE THREAD

**Issue by drujensen** [johnjansen/crystal-libraries-needed#17]

**Comment by elorest** [comment]
ICR exists. There's obviously some differences since crystal is a compiled language. I don't think it lets you use vim inside of it like pry.

**Comment by TheLonelyGhost** [comment]
Honestly, I'd be happy with it if I could view source code or documentation for a class/module/method. Being able to "cd" into a construct (whether class or instantiated object) is incredibly helpful when debugging.

**Comment by jwoertink** [comment]
I actually started working on a way to do this in ICR. The libreadline library actually has a lot of really useful features that would help us to achieve this. Also, somehow integrating lldb would as well. The issue with the readline is that it's lacking in crystal. I made an issue for one part https://github.com/crystal-lang/crystal/issues/4067 which is macOS related, but that prevents me from moving forward. 

I'd like to tackle this again if anyone wants to jump on board with me!

**Comment by elorest** [comment]
That would be cool. If I remember from the issue though they claim it's
related to the library for ReadLine in C rather than Crystal

**Comment by jwoertink** [comment]
Yeah, that's what they say; however, similar code runs in other languages. I don't think it's a readline issue, but I don't know the library well enough to be 100% sure.

**Comment by schovi** [comment]
@elorest For example Haskell is compiled too and there is really pretty interactive env for that. What is different from Crystal?

**Comment by citizen428** [comment]
@schovi GHC for Haskell has two modes, compiled and interpreted. `runghc` and `ghci` use the latter.

**Comment by schovi** [comment]
@citizen428 thanks for explanation!

**Comment by Sija** [comment]
@jwoertink You could try using [fancyline](https://github.com/Papierkorb/fancyline) instead of _libreadline_.

**Comment by veelenga** [comment]
AFAIK `readline` (or `fancyline`) allows you to only interrupt execution and read/evaluate an interactive command (aka repl). `binding.pry` in ruby is much more than just `readline`. Let's look at the example:

```ruby
require 'pry'

class A
end

def ruby_method1
end

def ruby_method2(a, b)
  binding.pry # <--- execution stopped here
end

ruby_method2 3, "a"
```

```sh
[1] pry(main)> a
=> 3
[2] pry(main)> b
=> "a"
[3] pry(main)> A.new
=> #<A:0x007f93e06910d8>
[4] pry(main)> ruby_method1
=> nil
[5] pry(main)> c
NameError: undefined local variable or method `c' for main:Object
from (pry):7:in `ruby_method2'
```

here `pry` knows the execution context it is running in, so it can show you the values of method args, create an instance of available class `A` and call a method. Also, it knows that there is no var `c`, for example. This is achieved by the [`Binding`](https://ruby-doc.org/core-2.2.0/Binding.html) class in ruby:

> Objects of class Binding encapsulate the execution context at some particular place in the code and retain this context for future use. The variables, methods, value of self, and possibly an iterator block that can be accessed in this context are all retained. Binding objects can be created using Kernel#binding, and are made available to the callback of Kernel#set_trace_func.

If we could have something similar in Crystal, it would be really easy to implement analog of `binding.pry`. Take a look at the first pry's release in ruby: https://github.com/pry/pry/blob/4fcf81b97601945945f43311532c164a93b44d7c/lib/pry.rb 

But creating `Binding` may look like a challenge for Crystal...

**Comment by Mic92** [comment]
I once built a proof of concept for C++ to make this possible: https://github.com/inspector-repl/inspector

**Comment by watzon** [comment]
I made a pure crystal library that doesn't rely on readline called [Term::Reader](https://github.com/crystal-term/reader) if anyone would like to use that for something like this. Not sure about the rest of the `pry` functionality though.

**Comment by Sija** [comment]
> I made a pure crystal library that doesn't rely on readline called [Term::Reader](https://github.com/crystal-term/reader) if anyone would like to use that for something like this. Not sure about the rest of the `pry` functionality though.

@watzon that's completely off-topic
